fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/data/RepaymentPeriod.java

```diff
@@ -0,0 +1,176 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.fineract.portfolio.loanaccount.loanschedule.data;
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+import org.apache.fineract.infrastructure.core.service.MathUtil;
+import org.apache.fineract.organisation.monetary.domain.Money;
+
+@ToString
+@EqualsAndHashCode(exclude = { "previous", "next" })
+public class RepaymentPeriod {
+
+    @ToString.Exclude
+    private final RepaymentPeriod previous;
```

Comment:
Are you sure you want to create a linked list from this? 

What's the reasoning behind the recursive structure?

Reply:
Yes, the algorithm is simpler and more performant if there is a reference to the previous reference.
We also store these repayment periods in an ArrayList to utilize List features.
When this list constructed then it has constant elements.
Unfortunately we can't use java's LinkedList implementation because it hides the important pointer references, so you can't use a real linked list.

Reply:
I'm not necessarily in agreement with your point - at least when it comes to performance.
Do you have measurements on how much "more performant" it is?

Let's say we save 10ms by using this recursive approach - even though from maintenance and debugging point of view it's adding a huge complexity - compared to an imperative approach which takes 20ms. We do multiple queries to the DB which takes let's say 300ms. That means with the imperative approach, we have 320ms execution time overall and with the recursive it's 310ms. We saved 3% of the overall executing while having an error-prone and hard to debug approach.

I'd go imperative, unless we have a super good reasoning - other than the recursive vs imperative performance difference.

Reply:
I am not sure i am following you guys. This is not a recursive or imperative approach. We are storing a pointer to the previous repayment period because there is recalculation logic which read some of the values from the previous repaymend period and this way it was super simple... nothing else... we are not using it to iterate on repayment periods or anything else...